### PR TITLE
[Testing] Reworked tests to not use HTML & shared functions

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

- Reworked tests to not use HTML & shared functions
- Added the new functions to `helper.psm1`

| Pipeline |
| - |
|[AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml)|

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: arm/.global/global.module.tests.ps1 [hunk 1]
```diff
@@ -279,67 +266,88 @@ Describe 'Readme tests' -Tag Readme {
...
+            $expectedParamCategories = $descriptions | ForEach-Object { $_.Split('.')[0] } | Select-Object -Unique # Get categories in template
+            $expectedParamCategoriesInOrder = $expectColumnsInOrder | Where-Object { $_ -in $expectedParamCategories } # add required ones in order
+            $expectedParamCategoriesInOrder += $expectedParamCategories | Where-Object { $_ -notin $expectColumnsInOrder } # add non-required ones after
+
+            $actualParamCategories = $readMeContent | Select-String -Pattern '^\*\*(.+) parameters\*\*$' -AllMatches | ForEach-Object { $_.Matches.Groups[1].Value } # get actual in readme
+
+            $actualParamCategories | Should -Be $expectedParamCategoriesInOrder
+        }
+
+        It '[<moduleFolderName>] parameter tables should contain only columns if they contain data, and if, in order: Parameter Name, Type, Default Value, Allowed Values, Description' -TestCases $readmeFolderTestCases {
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:292]: Trying with this
```suggestion
        It '[<moduleFolderName>] parameter tables should provide columns in the following order: Parameter Name, Type, Default Value, Allowed Values, Description. Each column should be present unless empty for all the rows.' -TestCases $readmeFolderTestCases {
```

File: arm/.global/global.module.tests.ps1 [hunk 2]
```diff
@@ -150,9 +156,17 @@ Describe 'Readme tests' -Tag Readme {
                 if (Test-Path (Join-Path $moduleFolderPath 'deploy.bicep')) {
                     $templateFilePath = Join-Path $moduleFolderPath 'deploy.bicep'
                     $templateContent = az bicep build --file $templateFilePath --stdout | ConvertFrom-Json -AsHashtable
+
+                    if (-not $templateContent) {
+                        throw "Unable to compile the deploy.bicep template's content. This can happen if there is an error in the template. Please check if you can run the command `$templateContent = az bicep build --file $templateFilePath --stdout | ConvertFrom-Json -AsHashtable`."
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:161]: This is thrown by multiple tests. Should we create a custom exception to be reused?

File: arm/.global/global.module.tests.ps1 [hunk 3]
```diff
@@ -150,9 +156,17 @@ Describe 'Readme tests' -Tag Readme {
...
+                    }
                 } elseIf (Test-Path (Join-Path $moduleFolderPath 'deploy.json')) {
                     $templateFilePath = Join-Path $moduleFolderPath 'deploy.json'
                     $templateContent = Get-Content $templateFilePath -Raw | ConvertFrom-Json -AsHashtable
+
+                    if (-not $templateContent) {
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:167]: Also this is thrown by multiple tests. Should we create a custom exception to be reused?

File: arm/.global/global.module.tests.ps1 [hunk 4]
```diff
@@ -184,52 +199,36 @@ Describe 'Readme tests' -Tag Readme {
         }
 
         It '[<moduleFolderName>] Readme.md file should contain the these titles in order: Resource Types, Parameters, Outputs' -TestCases $readmeFolderTestCases {
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:201]:
```suggestion
        It '[<moduleFolderName>] Readme.md file should contain these sections in order: Navigation, Resource Types, Parameters, Outputs, Deployment examples' -TestCases $readmeFolderTestCases {
```

File: arm/.global/global.module.tests.ps1 [hunk 5]
```diff
@@ -279,67 +266,88 @@ Describe 'Readme tests' -Tag Readme {
             $differentiatingItems.Count | Should -Be 0 -Because ("list of resources in the ReadMe's list [{0}] not in the template file should be empty" -f ($differentiatingItems -join ','))
         }
 
-        It '[<moduleFolderName>] parameters section should contain a table with these column names in order: Parameter Name, Type, Default Value, Possible values, Description' -TestCases $readmeFolderTestCases {
+        It '[<moduleFolderName>] should contain a parameter table for each parameter catagory in order: Required, Conditional, Optional, Generated' -TestCases $readmeFolderTestCases {
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:269]:
```suggestion
        It '[<moduleFolderName>] Parameters should contain a table for each existing parameter category in the following order: Required, Conditional, Optional, Generated' -TestCases $readmeFolderTestCases {
```

File: arm/.global/global.module.tests.ps1 [hunk 6]
```diff
@@ -279,67 +266,88 @@ Describe 'Readme tests' -Tag Readme {
...
+
+                $readMeCategoryIndex = $readMeContent | Select-String -Pattern "^\*\*$paramCategory parameters\*\*$" | ForEach-Object { $_.LineNumber }
+                $readmeCategoryColumns = ($readMeContent[$readMeCategoryIndex] -split '\|') | ForEach-Object { $_.Trim() } | Where-Object { -not [String]::IsNullOrEmpty($_) }
+
+                $readmeCategoryColumns | Should -Be $expectedColumnsInOrder
             }
-            ($ComparisonFlag -gt 2) | Should -Be $false
         }
 
         It '[<moduleFolderName>] parameters section should contain all parameters from the template file' -TestCases $readmeFolderTestCases {
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:327]:
```suggestion
        It '[<moduleFolderName>] Parameters section should contain all parameters from the template file' -TestCases $readmeFolderTestCases {
```

File: arm/.global/global.module.tests.ps1 [hunk 7]
```diff
@@ -243,30 +242,18 @@ Describe 'Readme tests' -Tag Readme {
         }
 
         It '[<moduleFolderName>] Resources section should not contain more resources as in the template file' -TestCases $readmeFolderTestCases {
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:244]:
```suggestion
        It '[<moduleFolderName>] Resources section should not contain more resources than the template file' -TestCases $readmeFolderTestCases {
```

File: arm/.global/global.module.tests.ps1 (hunk 2 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1:161]: Hey @eriqua, do you mean like a variable outside of all describe blocks that we then pass in via `$script:compilationException = '(...)'?

File: arm/.global/global.module.tests.ps1 (hunk 3 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1:167]: see other comment

File: arm/.global/global.module.tests.ps1 [hunk 8]
```diff
@@ -392,45 +392,34 @@ Describe 'Readme tests' -Tag Readme {
         }
 
         It '[<moduleFolderName>] Output section should contain all outputs defined in the template file' -TestCases $readmeFolderTestCases {
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1:394]: Agreed. We should test that

File: arm/.global/global.module.tests.ps1 [hunk 9]
```diff
@@ -155,20 +160,20 @@ Describe 'Readme tests' -Tag Readme {
...
                     $templateFilePath = Join-Path $moduleFolderPath 'deploy.json'
                     $templateContent = Get-Content $templateFilePath -Raw | ConvertFrom-Json -AsHashtable
 
                     if (-not $templateContent) {
-                        throw "Unable to load the deploy.json template's content. This can happen if there is an error in the template. Please check if you can run the command `Get-Content $templateFilePath -Raw | ConvertFrom-Json -AsHashtable`."
+                        throw ($jsonTemplateLoadFailedException -f $templateFilePath)
                     }
                 } else {
-                    throw "No template file found in folder [$moduleFolderPath]"
+                    throw ($noTemplateFoundException -f $moduleFolderPath)
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:176]:
```suggestion
                    throw ($templateNotFoundException -f $moduleFolderPath)
```

File: arm/.global/global.module.tests.ps1 [hunk 10]
```diff
@@ -462,20 +462,20 @@ Describe 'Deployment template tests' -Tag Template {
...
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:478]:
```suggestion
                    throw ($templateNotFoundException -f $moduleFolderPath)
```

File: arm/.global/global.module.tests.ps1 [hunk 11]
```diff
@@ -936,20 +936,20 @@ Describe "API version tests [All apiVersions in the template should be 'recent']
...
                 $templateFilePath = Join-Path $moduleFolderPath 'deploy.json'
                 $templateContent = Get-Content $templateFilePath -Raw | ConvertFrom-Json -AsHashtable
 
                 if (-not $templateContent) {
-                    throw "Unable to load the deploy.json template's content. This can happen if there is an error in the template. Please check if you can run the command `Get-Content $templateFilePath -Raw | ConvertFrom-Json -AsHashtable`."
+                    throw ($jsonTemplateLoadFailedException -f $templateFilePath)
                 }
             } else {
-                throw "No template file found in folder [$moduleFolderPath]"
+                throw ($noTemplateFoundException -f $moduleFolderPath)
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:952]:
```suggestion
                throw ($templateNotFoundException -f $moduleFolderPath)
```

File: arm/.global/global.module.tests.ps1 [hunk 12]
```diff
@@ -23,6 +23,11 @@ $script:enforcedTokenList = $enforcedTokenList
 # For runtime purposes, we cache the compiled template in a hashtable that uses a formatted relative module path as a key
 $script:convertedTemplates = @{}
 
+# Shared exception messages
+$script:bicepTemplateCompilationFailedException = "Unable to compile the deploy.bicep template's content. This can happen if there is an error in the template. Please check if you can run the command `az bicep build --file {0} --stdout | ConvertFrom-Json -AsHashtable`." # -f $templateFilePath
+$script:jsonTemplateLoadFailedException = "Unable to load the deploy.json template's content. This can happen if there is an error in the template. Please check if you can run the command `Get-Content {0} -Raw | ConvertFrom-Json -AsHashtable`." # -f $templateFilePath
+$script:noTemplateFoundException = 'No template file found in folder [{0}]' # -f $moduleFolderPath
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1]: done

File: arm/.global/global.module.tests.ps1 (hunk 3 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1:172]: done

File: arm/.global/global.module.tests.ps1 (hunk 9 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1]: done

File: arm/.global/global.module.tests.ps1 (hunk 10 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1]: done

File: arm/.global/global.module.tests.ps1 (hunk 11 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1]: done

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**